Fetching brs during approval

aaronk opened the pull request:
Improved fetching of BRs when approving.  Does a batch of promises at a time.

aaronk committed:
Fixed setState so that it returns a promise and response

aaronk committed:
remove console.log

aaronk committed:
clean up some error tracking

luthian commented on src/components/billingRecord/IFXBillingRecordList.vue:
```diff
@@ -181,16 +186,14 @@ export default {
-        console.log('the items now ', this.items)
         return br
       }
       console.log(`Billing record with id not found at item index ${index}`)
```
I assume this shouldn't happen but you want to log them somewhere?

aaronk replied:
I think it's a developer only thing.

luthian commented on src/components/billingRecord/IFXBillingRecordList.vue:
```diff
@@ -292,20 +295,24 @@ export default {
-            })
+      const promises = missing.map((i) => me.getFullBillingRecordByItemIndex(i))
+      while (promises.length) {
+        await new Promise(r => setTimeout(r, 500));
```
Why are you batching these promises up and waiting between each? Does the backend get overwhelmed or something? Or is there supposed to be a progress bar of some kind?

aaronk replied:
Mostly there would be 20-30.  A big batch of RC could be many hundreds.  Could be a bit much.

luthian replied:
You're firing them all off at once but then waiting in batches; that's the part I don't really understand. Do you want to show a progress indicator or something between waits? Or is the `sleep` there to allow user interaction?

luthian commented on src/components/billingRecord/IFXBillingRecordList.vue:
```diff
@@ -292,20 +295,24 @@ export default {
+      }
+      const toBeUpdated = []
+      this.items.forEach((it1) => {
+        items.forEach((it2) => {
```
~~Should this be `this.items` as above? Or is `items` defined somewhere?~~ Never mind.

luthian commented on src/components/billingRecord/IFXBillingRecordList.vue:
```diff
@@ -292,20 +295,24 @@ export default {
-          return this.$api.billingRecord.bulkUpdate(toBeUpdated, this.facility.applicationUsername)
+      }
+      const toBeUpdated = []
+      this.items.forEach((it1) => {
```
You might want to structure this differently to allow the loop to be broken out of once a match is found.
```suggestion
      items.forEach((it2) => {
        for (i = 0; i < this.items.length; i++) {
          if (this.items[i].id === it2.id) {
              const it1 = this.items[i]
              it1.billingRecordStates.push({ name: state, user: '', approvers: [], comment: '' })
             toBeUpdated.push(it1)
             break;
           }
         }
      }
```
Or something like that. The point is that `forEach` doesn't allow one to use `break`.

aaronk committed:
Update src/components/billingRecord/IFXBillingRecordList.vue

Co-authored-by: DDA <github@dantonio.net>

luthian commented on src/components/billingRecord/IFXBillingRecordList.vue:
```diff
@@ -292,20 +295,33 @@ export default {
+           }
+         }
+      }
+        items.forEach((it2) => {
```
I can't figure out how to do a multiline comment but you should remove the duplicated lines.

luthian committed:
Redo setState to pause after batches of requests for full billing record

- Only fetch record if `billingRecordState` isn’t already there
- Don’t bother updating the `items` list since approval will refresh anyway
